proof/refine/ARM/Tcb_R.thy

```diff
@@ -794,49 +796,541 @@ lemma threadSet_valid_objs_tcbPriority_update:
+  apply (wpsimp wp: tcbSchedEnqueue_all_invs_but_ct_not_inQ hoare_vcg_const_imp_lift
+                    rescheduleRequired_all_invs_but_ct_not_inQ hoare_vcg_all_lift)
+       apply (rule_tac Q="\<lambda>_. invs' and st_tcb_at' runnable' t" in hoare_strengthen_post[rotated])
+        apply (clarsimp simp: invs'_def valid_state'_def valid_pspace'_def comp_def)
+       apply (wpsimp wp: threadGet_wp threadSetPriority_invs' tcbSchedDequeue_not_queued)+
+  apply (drule invs_queues')
+  apply (fastforce simp: ready_qs_runnable_def valid_queues'_def inQ_def obj_at'_def)
+  done
+
+lemma runnable'_case_thread_state_If:
+  "(case rv of Structures_H.thread_state.Running \<Rightarrow> threadSetPriority_onRunning t x
+             | Structures_H.thread_state.Restart \<Rightarrow> threadSetPriority_onRunning t x
+             | _ \<Rightarrow> P) =
+   (if runnable' rv then threadSetPriority_onRunning t x else P)"
+  by (cases rv; clarsimp)
+
+ lemma setP_invs':
+   "\<lbrace>\<lambda>s. invs' s \<and> ready_qs_runnable s \<and> st_tcb_at' runnable' (ksCurThread s) s \<and> p \<le> maxPriority\<rbrace>
+    setPriority t p
+    \<lbrace>\<lambda>rv. invs'\<rbrace>"
+   apply (simp add: setPriority_def runnable'_case_thread_state_If)
+   apply (wpsimp wp: threadSetPriority_onRunning_invs' threadSetPriority_invs' reorderEp_invs'
+                     reorderNtfn_invs' gts_wp' hoare_vcg_all_lift hoare_vcg_const_imp_lift)
+   apply (drule invs_queues')
+   apply (fastforce simp: ready_qs_runnable_def valid_queues'_def inQ_def
+                          pred_tcb_at'_def obj_at'_def projectKO_eq)
+   done
+
+lemma tcb_ep_find_index_corres:
+  "corres (=) (tcb_at t and (\<lambda>s. \<forall>t \<in> set list. tcb_at t s) and K (n < length list))
+              (tcb_at' t and (\<lambda>s. \<forall>t \<in> set list. tcb_at' t s))
+              (tcb_ep_find_index t list n) (tcbEPFindIndex t list n)"
+  apply (rule corres_gen_asm')
+  apply (induct n)
+   apply (subst tcb_ep_find_index.simps)
+   apply (subst tcbEPFindIndex.simps)
+   apply (rule_tac r'="(=)" in corres_split)
+      apply (rule_tac r'="(=)" in corres_split)
+         apply (rule corres_if)
+           apply clarsimp
+          apply (rule corres_trivial)
+          apply clarsimp
+         apply (rule corres_trivial)
+         apply clarsimp
+        apply (rule threadget_corres)
+        apply (clarsimp simp: tcb_relation_def)
+       apply wpsimp
+      apply wpsimp
+     apply (rule threadget_corres)
+     apply (clarsimp simp: tcb_relation_def)
+    apply wpsimp
+   apply wpsimp
+  apply (subst tcb_ep_find_index.simps)
+  apply (subst tcbEPFindIndex.simps)
   apply (rule corres_guard_imp)
-    apply (rule sp_corres2)
-   apply (clarsimp simp: valid_sched_def valid_sched_action_def)
-  apply (clarsimp simp: invs'_def valid_state'_def sch_act_wf_weak)
+    apply (rule_tac r'="(=)" in corres_split)
+       apply (rule_tac r'="(=)" in corres_split)
+          apply (rule corres_if)
```

Comment:
suggestion: when clarsimp immediately solves the next goal, I like to move it to the previous command with a comma. E.g. `apply (rule foo, clarsimp)` I think it makes the proofs easier to read.